src/utils/contracts.ts

```diff
@@ -72,64 +112,184 @@ export function composableCowContract(
-        case "INTERFACE_NOT_SUPPORTED":
-        case "INVALID_FALLBACK_HANDLER":
-        case "INVALID_HANDLER":
-        case "SWAP_GUARD_RESTRICTED":
-          return { selector };
-        case "ORDER_NOT_VALID":
-        case "POLL_TRY_NEXT_BLOCK":
-        case "POLL_NEVER":
-          const message = ethers.utils.defaultAbiCoder.decode(
-            ["string"],
-            "0x" + errorData.slice(10) // trim off the selector
-          )[0];
-          return { selector, message };
-        case "POLL_TRY_AT_BLOCK":
-        case "POLL_TRY_AT_EPOCH":
-          const [blockNumberOrEpoch, msg] = ethers.utils.defaultAbiCoder.decode(
-            ["uint256", "string"],
-            "0x" + errorData.slice(10) // trim off the selector
-          );
-          return {
-            selector,
-            message: msg,
-            blockNumberOrEpoch: blockNumberOrEpoch.toNumber(),
-          };
-      }
-    } catch (err) {
-      // This can only return an error if the ABI decoding fails, which should never happen
-      // except for a bug in the smart contract code or the ABI encoding. In this case, we
-      // just bubble up the error.
-      throw err;
+    // Below here, the only throw that can happen is if the revert data contains a selector that is in the
+    // CUSTOM_ERROR_SELECTOR_MAP, but the it's parameters are not ABI-encoded correctly.
+
+    const selector = CUSTOM_ERROR_SELECTOR_MAP[rawSelector];
+    const fragment = ethers.utils.Fragment.fromString(
+      "error " + CUSTOM_ERROR_ABI_MAP[selector]
+    );
+    const iface = new ethers.utils.Interface([fragment]);
+
+    switch (selector) {
+      case CustomErrorSelectors.PROOF_NOT_AUTHED:
+      case CustomErrorSelectors.SINGLE_ORDER_NOT_AUTHED:
+      case CustomErrorSelectors.INTERFACE_NOT_SUPPORTED:
+      case CustomErrorSelectors.INVALID_FALLBACK_HANDLER:
+      case CustomErrorSelectors.INVALID_HANDLER:
+      case CustomErrorSelectors.SWAP_GUARD_RESTRICTED:
+        return { selector };
+      case CustomErrorSelectors.ORDER_NOT_VALID:
+      case CustomErrorSelectors.POLL_TRY_NEXT_BLOCK:
+      case CustomErrorSelectors.POLL_NEVER:
+        const [message] = iface.decodeErrorResult(fragment, revertData);
+        return { selector, message };
+      case CustomErrorSelectors.POLL_TRY_AT_BLOCK:
+      case CustomErrorSelectors.POLL_TRY_AT_EPOCH:
+        const [blockNumberOrEpoch, msg] = iface.decodeErrorResult(
+          fragment,
+          revertData
+        );
+
+        // It is reasonable to expect that the block number or epoch is bound by
+        // uint32. It is therefore safe to throw if the value is outside of that
+        // for javascript's number type.
+        if (blockNumberOrEpoch.gt(MAX_UINT32)) {
+          throw new Error("Block number or epoch out of bounds");
+        }
+
+        return {
+          selector,
+          message: msg as string,
+          blockNumberOrEpoch: (blockNumberOrEpoch as BigNumber).toNumber(),
```

Comment:
do we really need this hard casting of types?

Reply:
I moved the explicit hard casting to be on the same line as the `decode`. The hard casting is required because when ABI decoding in ethers v5, the return type is `Result`, which has the type:

```typescript
export interface Result extends ReadonlyArray<any> {
    readonly [key: string]: any;
}
```

IMO we should always hard cast from a decode if the decode is returning `Result`, as this increases the readability of the code, and enforces that Typescript catches more potential issues.